File: impls/lua-lukas/step1_read_print.lua
function READ(arg1)
  return arg1
end

function EVAL(arg1)
  return arg1
end

function PRINT(arg1)
  return arg1
end

function rep(arg1)
  return PRINT(EVAL(READ(arg1)))
end


run = true
while (run) do
  print "user> "
  input = io.stdin:read()
  if (input == nil)  then
    run = false
  else
    print(rep(input))
  end
end

File: impls/lukas-rust/step1_read_print.rs
mod printer;
mod reader;
mod types;

use crate::types::MalType;

fn main() {
    loop {
        println!("user> ");
        let mut line = String::new();
        let res = std::io::stdin().read_line(&mut line);

        match res {
            Ok(bytes) => {
                if bytes == 0 {
                    break;
                } else {
                    let trimmed_line = line.trim_end();
                    // let tokenized = reader::read_str(&trimmed_line);
                    // let (form, _) = reader::read_form(&tokenized);
                    // println!("{:?}", form);
                    let result: String = print(eval(read(&trimmed_line)));
                    println!("{}", result);
                }
            }
            Err(err) => {
                println!("Failed to read input, {}", err);
                break;
            }
        }
    }
}

fn read(s: &str) -> MalType {
    let reader = reader::read_str(s);
    reader::read_form(&reader).0
}

fn eval(s: MalType) -> MalType {
    s
}

fn print(mal: MalType) -> String {
    printer::pr_str(&mal)
}
